util/clustering/kmeans_clustering.py

```diff
@@ -0,0 +1,231 @@
+        parser.add_argument('--input-data-dir', type=str)   
+    
+    parser.add_argument('--analyze-clusters', type=int, default=3)   
+
+    args = parser.parse_args()
+
+    
+    file_dir = os.getcwd()
+    # # change the working directory to the path where the images are located
+    output_path = args.input_data_dir
+    os.chdir(output_path)
+
+    detection_train_images = []
+    # creates a ScandirIterator aliased as files
+    print('------------ getting image names ------------')
+    with os.scandir(output_path) as files:
+      # loops through each file in the directory
+        for file in files:
+            if file.name.endswith('.jpg'):
+              # adds only the image files to the detection_train_imagess list
+                detection_train_images.append(file.name)
+   
+    # print('get bucket info')
+    # bucket = "calvinandpogs-ee148"
+    # detection_train_images_path = "atrw/detection/train"
+    # s3 = boto3.resource("s3")
+    # s3_bucket = s3.Bucket(bucket)
+    # print('bucket info collected')
+
+    # # this list holds all the image filename
+    # detection_train_images = [f.key.split(detection_train_images_path + "/")[1] for f in s3_bucket.objects.filter(Prefix=detection_train_images_path).all()]
+    print(str(len(detection_train_images)) + ' images collected')
+
+        
+    model = VGG16()
+    model = Model(inputs = model.inputs, outputs = model.layers[-2].output)
+
+    def extract_features(file, model):
+        # load the image from the s3 bucket as a 224x224 array and convert from 
+        # 'PIL.Image.Image' to numpy array because the VGG model expects the images
+        # it receives to be 224x224 NumPy arrays.
+
+        # object = s3.Object(bucket, detection_train_images_path + '/' + file)
+        # img = load_img(BytesIO(object.get()['Body'].read()), target_size=(224,224))
+        img = load_img(file, target_size=(224,224))
+
+        img = np.array(img) 
+        # reshape the data for the model reshape(num_of_samples, dim 1, dim 2, channels)
+        reshaped_img = img.reshape(1,224,224,3) 
+        # prepare image for model
+        imgx = preprocess_input(reshaped_img)
+        # get the feature vector
+        features = model.predict(imgx, use_multiprocessing=True)
+        return features
+    
+    data = {}
+    # p = r"CHANGE TO A LOCATION TO SAVE FEATURE VECTORS"
+
+    # loop through each image in the dataset
+    # temporarily trying first 1000 images
+    for i in tqdm(range(1000)):
+        # try to extract the features and update the dictionary
+        feat = extract_features(detection_train_images[i],model)
+        data[detection_train_images[i]] = feat
+        
+        # if something fails, save the extracted features as a pickle file (optional)
+        # except:
+        #     with open(p,'wb') as file:
+        #         pickle.dump(data,file)
+            
+    
+    # get a list of the filenames
+    filenames = np.array(list(data.keys()))
```

Comment:
This is turning a list of strings into a numpy array for no reason (that I can see). This is unnecessary and doesn't really make sense, just leave it as  a python list.

Reply:
I think it's just because numpy arrays have better performance?